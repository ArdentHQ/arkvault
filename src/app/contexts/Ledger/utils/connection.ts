import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
