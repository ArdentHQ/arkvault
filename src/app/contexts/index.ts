export * from "./Configuration";
export * from "./Environment";
export * from "./Ledger";
export * from "./Navigation";
export * from "./Steps";
