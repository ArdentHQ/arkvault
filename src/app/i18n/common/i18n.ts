export const translations = {
	ACTION: "Action",
	ADAPTER: "Adapter",
	ADD: "Add",
	ADDRESS: "Address",
	ADDRESSES: "Addresses",
	ADDRESSING: "Addressing",
	ADDRESS_NAME: "Address Name",
	ADD_NEW: "Add New",
	ADVANCED_MODE: "Advanced Mode",
	AGE: "Age",
	AGREE_DO_NOT_SHOW_AGAIN: "I agree, do not show this message again",
	ALERT: {
		DANGER: "Error",
		HINT: "Help",
		INFO: "Information",
		SUCCESS: "Success",
		WARNING: "Warning",
	},
	ALL: "All",
	AMOUNT: "Amount",
	AMOUNT_PLACEHOLDER: "Enter an amount",
	AND: "and",
	APPEARANCE: "Appearance",
	APP_NAME: "<strong>ARK</strong>Vault",
	ARK: "ARK",
	ASSET: "Asset",
	ASSETS: "Assets",
	ASSET_TYPE: "Asset Type",
	AT: "at",
	AUTHOR: "Author",
	AVAILABLE: "Available",
	AVATAR: "Avatar",
	AWAITING_SIGNATURE: "Awaiting Signature",
	BACK: "Back",
	BACKGROUND: "Background",
	BACK_TO_WALLET: "Back to Wallet",
	BALANCE: "Balance",
	BEST: "Best",
	BIP39_LANGUAGE: "Passphrase Language (BIP39)",
	BLOCK: "Block",
	BRIDGECHAIN: "Bridgechain",
	BUSINESS: "Business",
	CANCEL: "Cancel",
	CATEGORY: "Category",
	CHANGE: "Change",
	CHANGED: "Changed",
	CHOOSE: "Choose",
	CLIPBOARD: {
		SUCCESS: "Copied!",
		TOOLTIP_TEXT: "Copy to clipboard",
	},
	CLOSE: "Close",
	COIN: "Coin",
	COMING_SOON: "Coming soon",
	COMPLETE: "Complete",
	COMPLETED: "Completed",
	CONFIRM: "Confirm",
	CONFIRMATION_MODAL: {
		DESCRIPTION: "Are you sure you wish to navigate away from this page without saving?",
		TITLE: "Confirmation",
	},
	CONTACT: "Contact",
	CONTACTS: "Contacts",
	CONTACT_US: "Contact Us",
	CONTINUE: "Continue",
	CONTRACT: "Contract",
	COPY: "Copy",
	COPY_ADDRESS: "Copy Address",
	COPY_ERROR: "Copy Error",
	COPY_ID: "Copy Transaction ID",
	COPY_IPFS: "Copy IPFS Hash",
	CREATE: "Create",
	CRYPTOASSET: "Cryptoasset",
	CRYPTOCURRENCY: "Cryptocurrency",
	CURRENCY: "Currency",
	CURRENT: "Current",
	DATE: "Date",
	DATETIME: {
		DAYS_AGO_one: "A day ago",
		DAYS_AGO_other: "{{count}} days ago",
		FEW_SECONDS_AGO: "A few seconds ago",
		HOURS_AGO_one: "An hour ago",
		HOURS_AGO_other: "{{count}} hours ago",
		MINUTES_AGO_one: "A minute ago",
		MINUTES_AGO_other: "{{count}} minutes ago",
		MINUTES_one: "{{count}} minute",
		MINUTES_other: "{{count}} minutes",
		MONTHS_AGO_one: "A month ago",
		MONTHS_AGO_other: "{{count}} months ago",
		WEEKS_AGO_one: "A week ago",
		WEEKS_AGO_other: "{{count}} weeks ago",
		YEARS_AGO_one: "A year ago",
		YEARS_AGO_other: "{{count}} years ago",
	},
	DEFAULT_CHOSEN_FEE: "Default fee type",
	DELEGATE: "Delegate",
	DELEGATES: "Delegates",
	DELEGATE_ADDRESS: "Delegate Address",
	DELEGATE_NAME: "Delegate Name",
	DELETE: "Delete",
	DESCRIPTION: "Description",
	DETAILS: "Details",
	DEVELOPMENT_NETWORKS: "Development Networks",
	DISABLE: "Disable",
	DISCLAIMER: "Disclaimer",
	DOCUMENTATION: "Documentation",
	DONE: "Done",
	DOWNLOAD: "Download",
	DOWNLOADED: "Downloaded",
	DOWNLOADING: "Downloading",
	DO_NOT_SHOW_AGAIN: "Do not show this message again",
	EDIT: "Edit",
	EDIT_item: "Edit {{item}}",
	EMPTY_RESULTS: {
		SUBTITLE: "No results found. Refine your search and try again.",
		TITLE: "No Results",
	},
	ENABLE: "Enable",
	ENABLE_ADVANCED_MODE: "Enable Advanced Mode to use this feature",
	ENCRYPTED_WIF: "Encrypted WIF",
	EPOCH: "Epoch",
	ERROR: "Error",
	ERRORS: {
		INVALID_IMAGE: "File is not an image",
		INVALID_URL: "The link could not be opened: '{{url}}' is not a valid URL",
		TOO_BIG: "File is bigger than 1 MB",
		UNKNOWN: "Unknown Error",
	},
	EXCHANGE: "Exchange",
	EXPIRATION: "Expiration",
	EXPLORER: "Explorer",
	EXPORT: "Export",
	FAILED_FETCH: "Failed to fetch {{name}}. Reason: {{reason}}",
	FAILED_UPDATE: "Failed to update {{name}}. Reason: {{reason}}",
	FAILED_UPDATE_CHECK: "Checking for updates failed.",
	FEE: "Fee",
	FEE_AVERAGE: "Avg.",
	FEE_FAST: "Fast",
	FEE_SLOW: "Slow",
	FETCH: "Fetch",
	FIAT: "Fiat",
	FIAT_AMOUNT: "Fiat Amount",
	FIAT_FEE: "Fiat Fee",
	FIAT_TOTAL: "Fiat Total",
	FIAT_VALUE: "Fiat Value",
	FINISH: "Finish",
	FORGED_ARK: "Forged ARK",
	FROM: "From",
	GO_TO_PORTFOLIO: "Go to Portfolio",
	GO_TO_WALLET: "Go to Wallet",
	HASH: "Hash",
	HEIGHT: "Height",
	HERE: "here",
	HIDE: "Hide",
	HIDE_WALLET_BUTTON_TEXT: "Hide text from wallet buttons",
	ID: "ID",
	IMPORT: "Import",
	IMPORTED: "Imported",
	INFO: "Info",
	INITIALIZING: "Initializing...",
	INPUT_ADDRESS: {
		VALIDATION: {
			ADDRESS_ALREADY_EXISTS: "Address {{address}} already exists",
			NOT_VALID: "The address is not valid",
		},
	},
	INPUT_NETWORK: {
		PLACEHOLDER: "Enter a cryptoasset name",
		VALIDATION: {
			NETWORK_NOT_FOUND: "No cryptoasset with this name found",
		},
	},
	INPUT_PASSPHRASE: {
		VALIDATION: {
			ADDRESS_ALREADY_EXISTS: "Address {{address}} already exists",
			MNEMONIC_NOT_MATCH_WALLET: "This mnemonic does not correspond to your wallet",
			PASSWORD_NOT_MATCH_WALLET: "This password does not correspond to your wallet",
			PRIVATE_KEY_NOT_MATCH_WALLET: "This private key does not correspond to your wallet",
			SECRET_NOT_MATCH_WALLET: "This secret does not correspond to your wallet",
			WIF_NOT_MATCH_WALLET: "This WIF does not correspond to your wallet",
		},
	},
	INPUT_PUBLIC_KEY: {
		VALIDATION: {
			PUBLIC_KEY_ALREADY_EXISTS: "Public Key {{publicKey}} already exists",
		},
	},
	INSTALL: "Install",
	INSTALL_ARKVAULT: "Install ARK Vault",
	INSTALL_ARKVAULT_STEP1: "1. Press the share button",
	INSTALL_ARKVAULT_STEP2: `2. Select "Add to Home Screen"`,
	INSTALL_ARKVAULT_STEP3: `3. Press the "Add" button`,
	IS_MARKET_CHART_ENABLED: "Price chart on the dashboard",
	I_ACCEPT: "I Accept",
	I_DECLINE: "I Decline",
	LANGUAGE: "Application Language",
	LAUNCH: "Launch",
	LEARN_MORE: "Learn more",
	LEDGER: "Ledger",
	LEDGER_COMPATIBILITY_ERROR: "ARK Vault requires the use of a chromium based browser when using a Ledger.",
	LEDGER_COMPATIBILITY_ERROR_LONG:
		"Your portfolio contains 1 or more Ledger wallets. ARK Vault requires the use of a chromium based browser such as Chrome, Brave or Edge when using a Ledger device.",
	LEDGER_CONNECTED: "{{device}} connected",
	LEDGER_DISCONNECTED: "{{device}} disconnected",
	LEDGER_WALLET: "This is a Ledger wallet",
	LINK: "Link",
	LOAD: "Load",
	LOADING: "Loading ...",
	LOADING_LEDGER: "Loading Ledger...",
	MANAGE_SERVERS: "Manage Servers",
	MAX: "Max",
	MEMO: "Memo",
	MEMO_SMARTBRIDGE: "Memo (SmartBridge)",
	MESSAGE: "Message",
	METHOD: "Method",
	MNEMONIC: "Mnemonic",
	MNEMONIC_TYPE: {
		BIP39: "Mnemonic",
		BIP44: "Mnemonic",
		BIP49: "Mnemonic (SegWit)",
		BIP84: "Mnemonic (Native SegWit)",
	},
	MORE_DETAILS: "More Details",
	MOST_POPULAR: "Most popular",
	MULTIPLE: "Multiple",
	MULTISIG: "Multisig",
	MULTISIGNATURE: "Multisignature",
	MY_ADDRESS: "My Address",
	MY_WALLET: "My Wallet",
	NAME: "Name",
	NETHASH: "Nethash",
	NETWORK: "Network",
	NETWORK_NAME: "This name is provided by the network",
	NETWORK_OR_NETHASH: "network or nethash",
	NETWORK_STATUS: "Network Status",
	NEW_DELEGATE: "New Delegate",
	NEXT: "Next",
	NO: "No",
	NONCE: "Nonce",
	NOT: "not",
	NOTICE_NO_WALLETS: "Create or import a wallet first",
	NOTIFICATIONS: {
		ACTIONS: {
			READ_CHANGELOG: "Read Changelog",
			UPDATE: "Update",
		},
		EMPTY: "You have no notifications at this time.",
		MARK_ALL_AS_READ: "Mark all as read",
		PLUGINS_TITLE: "Plugins",
		TITLE: "Notifications",
		TRANSACTIONS_TITLE: "Transactions",
	},
	NOT_AVAILABLE: "N/A",
	NO_OPTIONS: "No options",
	OF: "of",
	OFFICIAL: "Official",
	OLD_DELEGATE: "Old Delegate",
	OPEN: "Open",
	OPEN_IN_EXPLORER: "Open in Explorer",
	OPTIONAL: "optional",
	OTHER: "Other",
	PAGE: "Page",
	"PAGE_#": "Page {{page}}",
	"PAGE_#_OF_#": "Page {{page}} of {{total}}",
	PAGINATION: {
		ENTER_NUMBER: "Enter the page number",
	},
	PARTICIPANT: "Participant",
	PASSWORD: "Password",
	PASSWORDS: "Passwords",
	PEER: "Peer",
	PERIOD: "Period",
	PERIODS: {
		DAILY: "Daily",
	},
	PERMISSIONS: "Permissions",
	PORTFOLIO: "Portfolio",
	PREV: "Prev",
	PREVIOUS: "Previous",
	PRIVATE_KEY: "Private Key",
	PROFILE: "Profile",
	PROFILE_SYNC_COMPLETED: "Data updated successfully",
	PROFILE_SYNC_FAILED:
		"There has been a network problem on <NetworkNames/> while trying to retrieve your wallet information and as a result, your data could be out of sync or incorrect. Click <RetryLink/> to retry.",
	PROFILE_SYNC_STARTED: "Synchronizing your wallets data. Please wait ...",
	PUBLIC_KEY: "Public Key",
	PUBLIC_NETWORKS: "Public Networks",
	QR_CODE: "QR Code",
	QR_CODE_HELP_TEXT: "The QR-code is updated automatically, you do not need to press anything.",
	QR_SHORT: "QR",
	QUANTITY: "Quantity",
	RANK: "Rank",
	RATE: "Rate",
	RATING: "Rating",
	RECIPIENT: "Recipient",
	RECIPIENTS: "Recipients",
	"RECIPIENT_#": "Recipient #{{count}}",
	REFRESH: "Refresh",
	REGISTER: "Register",
	REGISTRATIONS: "Registrations",
	RELOAD: "Reload",
	REMOVE: "Remove",
	RENAME: "Rename",
	REPEAT: "Repeat",
	REPORT: "Report",
	RESET: "Reset",
	RESET_FILTERS: "Reset Filters",
	RESIGN: "Resign",
	RESIGNED: "Resigned",
	RETRY: "Retry",
	SAVE: "Save",
	SAVE_FILE: {
		ERROR: "Could not save file: {{error}}",
		SUCCESS: "<strong>{{filePath}}</strong> has been saved successfully",
	},
	SAVE_FINISH: "Save & Finish",
	SEARCH: "Search",
	SEARCH_BAR: {
		FIND_IT: "Find it",
		PLACEHOLDER: "What would you like to search for?",
	},
	SECOND_MNEMONIC: "2nd Mnemonic",
	SECOND_SECRET: "2nd Secret",
	SECOND_SIGNATURE: "Second Signature",
	SECRET: "Secret",
	SECURITY: "Security",
	SELECT: "Select",
	SELECTED: "Selected",
	SELECT_ALL: "Select All",
	SELECT_A_PROFILE: "You should select a profile to access this URL",
	SELECT_A_WALLET: "You should select a wallet to deeplink transactions",
	SELECT_OPTION: "Select {{option}}",
	SEND: "Send",
	SENDER: "Sender",
	SENDER_ADDRESS: "Sender address",
	SERVER_STATUS: {
		DOWNGRADED:
			"One or more networks are experiencing downgraded performance. See server settings for additional information.",
		HEALTHY: "All servers are healthy.",
		UNAVAILABLE: "One or more networks are unavailable. See server settings for additional information.",
	},
	SETTINGS: "Settings",
	SHOW: "Show",
	SHOWING_RESULTS: "Showing {{count}} results",
	SHOW_ALL: "Show All",
	SIGN: "Sign",
	SIGNATORY: "Signatory",
	SIGNATURE: "Signature",
	SIGNED: "Signed",
	SIGN_IN: "Sign In",
	SIGN_OUT: "Sign Out",
	SIZE: "Size",
	SKIP: "Skip",
	SORT_BY: "Sort by",
	STARRED: "Starred",
	START: "Start",
	STATUS: "Status",
	SUBMIT: "Submit",
	SUCCESS: "Success",
	SUPPORT: "Support",
	SYMBOL: "Symbol",
	TEST_NETWORK: "Test Network",
	THEME: "Theme",
	TIME: "Time",
	TIMESTAMP: "Timestamp",
	TIME_FORMAT: "Time format",
	TITLE: "Title",
	TO: "To",
	TOKEN: "Token",
	TOTAL: "Total",
	TOTAL_BALANCE: "Total Balance",
	TOTAL_CURRENCY: "Total Currency",
	TRANSACTION: "Transaction",
	TRANSACTION_HISTORY: "Transaction History",
	TRANSACTION_SUMMARY: "Transaction Summary",
	TRANSACTION_TYPE: "Transaction Type",
	TRANSFER: "Transfer",
	TX_ID: "Tx ID",
	TYPE: "Type",
	UNSELECT: "Unselect",
	UNSELECTED: "Unselected",
	UNSELECT_ALL: "Unselect All",
	UNVOTE: "Unvote",
	UPDATE: "Update",
	UPDATE_LATER: "Update Later",
	UPDATE_NOW: "Update Now",
	UPDATING: "Updating...",
	URL: "URL",
	VALIDATING_URI: "Validating URI. Please wait ...",
	VALIDATION: {
		CONFIRM_PASSWORD_REQUIRED: "Password confirmation is required",
		EXISTS: "{{field}} already exists",
		FIELD_INVALID: "{{field}} invalid",
		FIELD_REQUIRED: "{{field}} required",
		HOST_FORMAT: "The host format is invalid",
		ILLEGAL_CHARACTERS: "The following characters are not allowed: {{- characters}}",
		IS_REQUIRED: "is required",
		MAX_LENGTH: "{{field}} should have at most {{maxLength}} characters",
		MIN: "{{field}} should be least {{min}}",
		MIN_LENGTH: "{{field}} should have at least {{minLength}} characters",
		MIN_RECIPIENTS: "At least one recipient is required",
		OPTIONAL: "This field is optional",
		PASSWORD_LEAKED: "Please change your password, it has been found in a previous breach",
		PASSWORD_MISMATCH: "Passwords do not match",
		PASSWORD_RULES: {
			LENGTH: "8 characters minimum",
			LOWERCASE: "One lowercase character",
			NEW: "Different from current password",
			NUMBER: "One number",
			SYMBOL: "One special character",
			UNCOMPROMISED: "No match in leaked database",
			UPPERCASE: "One uppercase character",
		},
		PASSWORD_SAME_AS_OLD: "New password cannot be old password",
		PASSWORD_WEAK:
			"The password must contain at least 1 lowercase character, 1 uppercase character, 1 numeric character, 1 special character and must be 8 characters or longer",
		PORT_FORMAT: "The port format is invalid",
		RECIPIENT_INVALID: "Recipient address is invalid",
		SUBJECT_MISMATCH: "The {{subject}} do not match",
	},
	VALUE: "Value",
	VERIFIED: "Verified Address: {{value}}",
	VERSION: "Version",
	VIEW: "View",
	VIEW_ALL: "View All",
	VIEW_MORE: "View More",
	VOTE: "Vote",
	VOTES: "Votes",
	VOTING: "Voting",
	WAITING: "Waiting",
	WALLET: "Wallet",
	WALLETS: "Wallets",
	WALLET_ADDRESS: "Wallet Address",
	WALLET_TYPE: "Wallet Type",
	WARNING: "Warning",
	WEBSITE: "Website",
	WELCOME: "Welcome",
	WIF: "WIF",
	WILL: "will",
	WORD: "Word",
	X_OF_Y: "{{0}} of {{1}}",
	YES: "Yes",
	YOU: "You",
	YOUR_BALANCE: "Your Balance",
	YOUR_PASSPHRASE: "Your Passphrase",
};
