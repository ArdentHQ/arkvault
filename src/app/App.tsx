import "focus-visible";

import React, { useRef } from "react";
import { ErrorBoundary } from "react-error-boundary";
import { I18nextProvider } from "react-i18next";
import { CacheProvider } from "@emotion/react";

import createCache from "@emotion/cache";
import { ConfigurationProvider, EnvironmentProvider, LedgerProvider, NavigationProvider } from "./contexts";
import { i18n as index18n } from "./i18n";
import { ZendeskProvider } from "./contexts/Zendesk";
import { ExchangeProvider } from "@/domains/exchange/contexts/Exchange";
import { AppRouter, GlobalStyles, Main } from "@/app/App.blocks";
import { initializeEnvironment } from "@/utils/environment";
import { ApplicationError } from "@/domains/error/pages";

export const App: React.VFC = () => {
	/**
	 * Ensure that the Environment object will not be recreated when the state changes,
	 * as the data is stored in memory by the `DataRepository`.
	 */
	const environment = useRef(initializeEnvironment());

	const cache = createCache({ key: "emotion-cache" });
	cache.compat = true;

	return (
		<ZendeskProvider>
			<I18nextProvider i18n={index18n}>
				<EnvironmentProvider env={environment.current}>
					<ConfigurationProvider defaultConfiguration={{ profileIsSyncingExchangeRates: true }}>
						<NavigationProvider>
							<ExchangeProvider>
								<ErrorBoundary FallbackComponent={ApplicationError}>
									<LedgerProvider>
										<CacheProvider value={cache}>
											<AppRouter>
												<GlobalStyles />
												<Main />
											</AppRouter>
										</CacheProvider>
									</LedgerProvider>
								</ErrorBoundary>
							</ExchangeProvider>
						</NavigationProvider>
					</ConfigurationProvider>
				</EnvironmentProvider>
			</I18nextProvider>
		</ZendeskProvider>
	);
};
