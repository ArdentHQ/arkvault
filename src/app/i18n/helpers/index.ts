import { translations as CONTACTS } from "@/domains/contact/i18n";
import { translations as DASHBOARD } from "@/domains/dashboard/i18n";
import { translations as ERROR } from "@/domains/error/i18n";
import { translations as EXCHANGE } from "@/domains/exchange/i18n";
import { translations as MESSAGE } from "@/domains/message/i18n";
import { translations as PROFILE } from "@/domains/profile/i18n";
import { translations as SETTINGS } from "@/domains/setting/i18n";
import { translations as TRANSACTION } from "@/domains/transaction/i18n";
import { translations as VOTE } from "@/domains/vote/i18n";
import { translations as WALLETS } from "@/domains/wallet/i18n";
import { translations as COMMON } from "@/app/i18n/common/i18n";

export const buildTranslations = () => ({
	COMMON,
	CONTACTS,
	DASHBOARD,
	ERROR,
	EXCHANGE,
	MESSAGE,
	PROFILE,
	SETTINGS,
	TRANSACTION,
	VOTE,
	WALLETS,
});
