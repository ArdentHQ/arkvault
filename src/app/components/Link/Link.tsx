import cn from "classnames";
import React from "react";
import { useTranslation } from "react-i18next";
import { Link as RouterLink, LinkProps } from "react-router-dom";
import tw, { styled } from "twin.macro";

import { Icon } from "@/app/components/Icon";
import { Tooltip } from "@/app/components/Tooltip";
import { toasts } from "@/app/services";
import { useLink } from "@/app/hooks/use-link";

const AnchorStyled = styled.a<{ isDisabled?: boolean }>(({ isDisabled }) => {
	const styles = [
		tw`relative space-x-1 font-semibold`,
		tw`transition-colors`,
		tw`cursor-pointer no-underline`,
		tw`focus:outline-none`,
	];

	if (isDisabled) {
		styles.push(tw`text-theme-secondary-text cursor-not-allowed`);
	} else {
		styles.push(tw`text-theme-primary-600 hover:text-theme-primary-700 active:text-theme-primary-400`);
	}

	return styles;
});

const Content = styled.span<{ isDisabled?: boolean }>(({ isDisabled }) => {
	const styles = [
		tw`break-all border-b border-transparent`,
		tw`[transition-property:color,_border-color]`,
		tw`[transition-duration:200ms,_350ms]`,
		tw`[transition-delay:0s, _100ms]`,
	];

	if (!isDisabled) {
		styles.push(tw`group-hover:border-current`);
	}

	return styles;
});

type AnchorProperties = {
	isDisabled?: boolean;
	isExternal?: boolean;
	showExternalIcon?: boolean;
} & React.AnchorHTMLAttributes<any>;

const Anchor = React.forwardRef<HTMLAnchorElement, AnchorProperties>(
	(
		{ isDisabled, isExternal, showExternalIcon, onClick, href, children, rel, className }: AnchorProperties,
		reference,
	) => (
		<AnchorStyled
			className={cn("ring-focus group inline-block", className)}
			data-testid="Link"
			rel={isExternal ? "noopener noreferrer" : rel}
			ref={reference}
			onClick={onClick}
			href={href}
			data-ring-focus-margin="-m-1"
			isDisabled={isDisabled}
		>
			<Content isDisabled={isDisabled}>{children}</Content>
			{isExternal && showExternalIcon && (
				<Icon
					data-testid="Link__external"
					name="ArrowExternal"
					className={cn("mb-[3px] shrink-0 align-middle duration-200", { "inline-block text-sm": children })}
				/>
			)}
		</AnchorStyled>
	),
);

Anchor.displayName = "Anchor";

type Properties = {
	isDisabled?: boolean;
	isExternal?: boolean;
	children?: React.ReactNode;
	tooltip?: string;
	showExternalIcon?: boolean;
} & Omit<LinkProps, "referrerPolicy">;

export const Link = ({
	className,
	tooltip,
	isDisabled = false,
	isExternal = false,
	showExternalIcon = true,
	...properties
}: Properties) => {
	const { t } = useTranslation();
	const { openExternal } = useLink();

	return (
		<Tooltip content={tooltip} disabled={!tooltip}>
			{isExternal ? (
				<Anchor
					className={className}
					onClick={(event) => {
						event.stopPropagation();
						event.preventDefault();

						if (!isDisabled) {
							try {
								openExternal(properties.to.toString());
							} catch {
								toasts.error(t("COMMON.ERRORS.INVALID_URL", { url: properties.to }));
							}
						}
					}}
					href={properties.to as string}
					showExternalIcon={showExternalIcon}
					isDisabled={isDisabled}
					isExternal
				>
					{properties.children}
				</Anchor>
			) : (
				<></>
				// <RouterLink component={Anchor} {...properties} />
			)}
		</Tooltip>
	);
};
