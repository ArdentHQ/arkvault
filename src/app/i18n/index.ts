import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// Helpers
import { buildTranslations } from "./helpers";

const defaultNS = "translation";

const resources = {
	en: {
		translation: buildTranslations(),
	},
};

i18n.use(initReactI18next).init({
	defaultNS,
	lng: "en",
	ns: [defaultNS],
	resources,
});

export { defaultNS, resources };

export { default as i18n } from "i18next";
