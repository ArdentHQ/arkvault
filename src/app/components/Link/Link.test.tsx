import userEvent from "@testing-library/user-event";
import React from "react";

import { Link } from "./Link";
import { buildTranslations } from "@/app/i18n/helpers";
import { toasts } from "@/app/services";
import { render, screen } from "@/utils/testing-library";

const translations = buildTranslations();

describe("Link", () => {
	it("should render", () => {
		const { asFragment } = render(<Link to="/test">Test</Link>);

		expect(screen.getByTestId("Link")).toHaveTextContent("Test");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render external", () => {
		render(
			<Link to="https://arkvault.io/" isExternal>
				ARK.io
			</Link>,
		);

		expect(screen.getByTestId("Link")).toHaveAttribute("rel", "noopener noreferrer");
		expect(screen.getByTestId("Link__external")).toBeInTheDocument();
	});

	it("should render disabled", () => {
		render(
			<Link to="https://arkvault.io/" isExternal isDisabled>
				ARK.io
			</Link>,
		);

		expect(screen.getByTestId("Link")).toHaveAttribute("rel", "noopener noreferrer");
		expect(screen.getByTestId("Link__external")).toBeInTheDocument();
	});

	it("should do nothing on click when disabled", () => {
		const windowSpy = jest.spyOn(window, "open").mockImplementation();
		const externalLink = "https://arkvault.io/";

		const { asFragment } = render(
			<Link to={externalLink} isExternal isDisabled>
				ARK.io
			</Link>,
		);

		expect(screen.getByTestId("Link")).toHaveAttribute("rel", "noopener noreferrer");
		expect(screen.getByTestId("Link__external")).toBeInTheDocument();

		userEvent.click(screen.getByTestId("Link"));

		expect(windowSpy).not.toHaveBeenCalledWith(externalLink, "_blank");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render external without children", () => {
		const { asFragment } = render(<Link to="https://arkvault.io" isExternal />);

		expect(screen.getByTestId("Link__external")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should open an external link", () => {
		const windowSpy = jest.spyOn(window, "open").mockImplementation();

		const externalLink = "https://arkvault.io/";

		const { asFragment } = render(<Link to={externalLink} isExternal />);

		userEvent.click(screen.getByTestId("Link"));

		expect(windowSpy).toHaveBeenCalledWith(externalLink, "_blank");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should show a toast when trying to open an invalid external link", () => {
		const externalLink = "invalid-url";

		const toastSpy = jest.spyOn(toasts, "error");

		const { asFragment } = render(<Link to={externalLink} isExternal />);

		userEvent.click(screen.getByTestId("Link"));

		expect(toastSpy).toHaveBeenCalledWith(translations.COMMON.ERRORS.INVALID_URL.replace("{{url}}", "invalid-url"));
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with tooltip", () => {
		const { asFragment, baseElement } = render(
			<Link to="/test" tooltip="Custom Tooltip">
				Test
			</Link>,
		);
		const link = screen.getByTestId("Link");

		userEvent.hover(link);

		expect(baseElement).toHaveTextContent("Custom Tooltip");

		userEvent.click(link);

		expect(asFragment()).toMatchSnapshot();
	});
});
