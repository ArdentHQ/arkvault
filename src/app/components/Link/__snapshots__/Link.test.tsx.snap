// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > should do nothing on click when disabled 1`] = `
<DocumentFragment>
  <a
    class="ring-focus group inline-block css-1flyqpm"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="https://app.arkvault.io/"
    rel="noopener noreferrer"
  >
    <span
      class="css-1cv5b0m"
    >
      app.arkvault.io
    </span>
    <div
      class="mb-[3px] shrink-0 align-middle duration-200 inline-block text-sm css-v0ob3f"
      data-testid="Link__external"
      height="16"
      width="16"
    >
      <svg
        id="arrow-external"
        viewBox="0 0 20 20"
        x="0"
        xml:space="preserve"
        xmlns="http://www.w3.org/2000/svg"
        y="0"
      >
        <path
          d="M18.999 6.43V1h-5.43m5.43 0L7.033 12.966M9.749 4h-7.5c-.625 0-1.25.625-1.25 1.25v12.5c0 .625.625 1.25 1.25 1.25h12.5c.625 0 1.25-.625 1.25-1.25v-7.5"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </a>
</DocumentFragment>
`;

exports[`Link > should render 1`] = `
<DocumentFragment>
  <a
    class="ring-focus group inline-block css-7nwwrh"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="#/test"
  >
    <span
      class="css-170792"
    >
      Test
    </span>
  </a>
</DocumentFragment>
`;

exports[`Link > should render external without children 1`] = `
<DocumentFragment>
  <a
    class="ring-focus group inline-block css-7nwwrh"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="https://app.arkvault.io"
    rel="noopener noreferrer"
  >
    <span
      class="css-170792"
    />
    <div
      class="mb-[3px] shrink-0 align-middle duration-200 css-v0ob3f"
      data-testid="Link__external"
      height="16"
      width="16"
    >
      <svg
        id="arrow-external"
        viewBox="0 0 20 20"
        x="0"
        xml:space="preserve"
        xmlns="http://www.w3.org/2000/svg"
        y="0"
      >
        <path
          d="M18.999 6.43V1h-5.43m5.43 0L7.033 12.966M9.749 4h-7.5c-.625 0-1.25.625-1.25 1.25v12.5c0 .625.625 1.25 1.25 1.25h12.5c.625 0 1.25-.625 1.25-1.25v-7.5"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </a>
</DocumentFragment>
`;
